Update to BYSETPOS support

nehresma opened the pull request:
In August 2016, Nicolas Marlier added BYSETPOS support (#349).  This PR updates that PR with a few small changes to run in modern Ruby (use Integer instead of Fixnum) and a more modern rspec.

Commit:
Support BYSETPOS for MONTHLY AND YEARLY freq

nehresma committed:
Modernize BYSETPOS commit

A few small updates to Nicolas Marlier's BYSETPOS support added in
PR #349

davidstosik commented:
I can confirm this works as expected, this is fantastic! 👍

davidstosik commented on .gitignore:
```diff
@@ -8,5 +8,8 @@
 
+# rubymine
+.idea
+
```
Is that necessary? It looks like an accidental addition that is not part of this PR.

davidstosik commented on lib/ice_cube/time_util.rb:
```diff
@@ -198,6 +200,36 @@ def self.which_occurrence_in_month(time, wday)
 
+    # Use Activesupport CoreExt functions to manipulate time
+    def self.start_of_month time
+      time.beginning_of_month
```
Why not reuse the same name?

davidstosik commented on lib/ice_cube/time_util.rb:
```diff
@@ -198,6 +200,36 @@ def self.which_occurrence_in_month(time, wday)
+
+    # Use Activesupport CoreExt functions to manipulate time
+    def self.start_of_year time
+      time.beginning_of_year
```
Why not reuse the same name?

davidstosik commented on lib/ice_cube/time_util.rb:
```diff
@@ -198,6 +200,36 @@ def self.which_occurrence_in_month(time, wday)
+
+    # Use Activesupport CoreExt functions to manipulate time
+    def self.previous_month time
+      time - 1.month
```
I think `time.last_month` exists?

davidstosik commented on lib/ice_cube/time_util.rb:
```diff
@@ -198,6 +200,36 @@ def self.which_occurrence_in_month(time, wday)
+
+    # Use Activesupport CoreExt functions to manipulate time
+    def self.previous_year time
+      time - 1.year
```
I think `time.last_year` exists?

davidstosik commented on lib/ice_cube/time_util.rb:
```diff
@@ -198,6 +200,36 @@ def self.which_occurrence_in_month(time, wday)
+      time - 1.month
+    end
+
+    # Use Activesupport CoreExt functions to manipulate time
```
I don't understand the point of adding those class methods to `TimeUtils`.
Wouldn't it be easier to just call `time.whatever` instead of `TimeUtils.whatever(time)` wherever we need those methods?

davidstosik commented on lib/ice_cube/validations/yearly_by_set_pos.rb:
```diff
@@ -0,0 +1,89 @@
+        end
+
+
+
```
[Nitpick] Too many empty lines?

davidstosik commented on spec/examples/by_set_pos_spec.rb:
```diff
@@ -0,0 +1,29 @@
+      schedule = IceCube::Schedule.from_ical "RRULE:FREQ=MONTHLY;COUNT=4;BYDAY=WE;BYSETPOS=4"
+      schedule.start_time = Time.new(2015, 5, 28, 12, 0, 0)
+      expect(schedule.occurrences_between(Time.new(2015, 01, 01), Time.new(2017, 01, 01))).to eq([
+                                                                                                     Time.new(2015,6,24,12,0,0),
```
[Nitpick] Any way to avoid such long lines?

davidstosik commented on lib/ice_cube/validations/monthly_by_set_pos.rb:
```diff
@@ -0,0 +1,85 @@
+  module Validations::MonthlyBySetPos
+
+    def by_set_pos(*by_set_pos)
+      return by_set_pos([by_set_pos]) if by_set_pos.is_a?(Integer)
```
As you used the splat operator in the method's signature, I don't think it is possible for `by_set_pos` to be an `Integer`. This line thus seems unnecessary.

davidstosik commented on lib/ice_cube/validations/monthly_by_set_pos.rb:
```diff
@@ -0,0 +1,85 @@
+      return by_set_pos([by_set_pos]) if by_set_pos.is_a?(Integer)
+
+      unless by_set_pos.nil? || by_set_pos.is_a?(Array)
+        raise ArgumentError, "Expecting Array or nil value for count, got #{by_set_pos.inspect}"
```
This validation seems unnecessary as well. `by_set_pos` **is always** an `Array`.

davidstosik commented on lib/ice_cube/validations/monthly_by_set_pos.rb:
```diff
@@ -0,0 +1,85 @@
+      end
+      by_set_pos.flatten!
+      by_set_pos.each do |set_pos|
+        unless (set_pos >= -366 && set_pos <= -1) ||
```
How about `(-366..366).include(set_post) && set_pos != 0`?

davidstosik commented on lib/ice_cube/validations/monthly_by_set_pos.rb:
```diff
@@ -0,0 +1,85 @@
+      by_set_pos.each do |set_pos|
+        unless (set_pos >= -366 && set_pos <= -1) ||
+            (set_pos <= 366 && set_pos >= 1)
+          raise ArgumentError, "Expecting number in [-366, -1] or [1, 366], got #{set_pos} (#{by_set_pos})"
```
Should this error display only the invalid `set_pos`, instead of the complete `by_set_post` list?

davidstosik commented on lib/ice_cube/validations/monthly_by_set_pos.rb:
```diff
@@ -0,0 +1,85 @@
+      end
+
+      @by_set_pos = by_set_pos
+      replace_validations_for(:by_set_pos, by_set_pos && [Validation.new(by_set_pos, self)])
```
If I understand correctly, `by_set_pos` cannot be `nil` or `false` here (it's an `Array`), in which case `by_set_pos && [Validation.new(by_set_pos, self)]` is equivalent to its right-side operand, `[Validation.new(by_set_pos, self)]`.

davidstosik commented on lib/ice_cube/validations/yearly_by_set_pos.rb:
```diff
@@ -0,0 +1,89 @@
+module IceCube
+
+  module Validations::YearlyBySetPos
```
Most of my comments in `MonthlyBySetPos` might apply here as well.

davidstosik reviewed (commented):
I left comments to see if we can improve the code of this PR.
I'll try implementing them myself and link a diff here a bit later.

nehresma commented:
Hello @davidstosik, @NicolasMarlier provided this original PR.  I simply updated it to work with modern versions of Ruby.

dimerman commented:
@nehresma what's the status of this PR? I'd very much like to use the `setpos` functionality.

nehresma commented:
@dimerman It has been merged but I do not believe a new version of the gem has been released.  I've pointed my Gemfile to the latest git ref on master for the time being.  Hopefully someone will have a few minutes to cut a new gem release at some point so we wont have to point to specific refs.

nehresma commented:
@dimerman Err, no -- I'm an idiot.  I just realized it has NOT been merged.  I checked my Gemfile and I'm pointing to the ref with BYSETPOS support out on my fork -- not master on seejohnrun/ice_cube.

Sorry for the confusion.  I should have double checked before trusting my memory.

dimerman commented:
@nehresma no worries - I've been pointing my gemfile to your branch/repo. thanks for that!

However, I found the following case is not working right:
```
rule_a=FREQ=MONTHLY;COUNT=12;BYDAY=MO,TU,WE,TH,FR;BYSETPOS=-1`
rule_b=FREQ=MONTHLY;COUNT=12;BYDAY=MO,TU,WE,TH,FR;BYMONTHDAY=13,14,15;BYSETPOS=-1
```

when I create a `IceCube::Schedule` and add each rule individually, the occurrences are accurate. But if I create a schedule with both these rules, I get 18 occurrences instead of the expected 24.
I thought you might want to know. 🙌

mattbooks commented:
What's still holding this back — the PR feedback seems pretty minor? If that gets cleaned up can we get this merged?

mattbooks commented:
@seejohnrun It seems like several people need this and are using forks to get around it being missing. I'd be more than happy to help clean this up to avoid using a fork myself if it is otherwise merge-able.

ovinix commented on lib/ice_cube/validations/monthly_by_set_pos.rb:
```diff
@@ -0,0 +1,85 @@
+
+
+      def build_s(builder)
+        builder.piece(:by_set_pos) << by_set_pos
```
Hello @nehresma 

Looks like this should be improved:
```
> IceCube::Rule.monthly(2).day(1,2,3,4,5).by_set_pos(2).to_s
"Every 2 months on Weekdays [2]"
```

pacso commented:
@nehresma / @mattbooks / @ovinix - If you can get this PR cleaned up and resolve the feedback and conflicts with `master`, I'll give it a proper review.

Naren1997 commented:
Hello, looks like the gem already supporting what BYTSETPOS does but in different ical format. Example: If we want recurring for every 2 months 3rd Wednesday, ical format w.r.t the BYSETPOS is **FREQ=MONTHLY;INTERVAL=1;BYDAY=MO;BYSETPOST=3** which is same as **FREQ=MONTHLY;INTERVAL=1;BYDAY=3MO**, the latter one is already supported by the gem, so not sure why we still wants this PR.

davidstosik commented:
> not sure why we still wants this PR

I don’t use the gem anymore, but you might want to consider that some people use the gem to parse strings produced by another piece of software they don’t have control on, and this would require the whole specification to be supported. 😉

jhubert commented:
@avit Anything we can do to help get this merged?

dimerman commented:
@davidstosik what gem do you use instead?

davidstosik commented:
Nothing. 😬 
My need to handle schedules has vanished. 🙇🏻‍♂️

nehresma committed:
address the spec DST sensitivity in .to_yaml round trips

nehresma committed:
update PR from feedback

rebased against master -- its been 4 years

nehresma commented:
It's been over 4 years since I created this PR and over 7 years since Nicolas Marlier's initial commit with this support.  I have rebased it against current master and I've applied changes per the feedback received above.

If someone has time to review and merge, that would be most appreciated.  BYSETPOS is part of RFC 5545 and it would be great if this gem could finally get support merged in rather than just the "noop" that it currently is (https://github.com/ice-cube-ruby/ice_cube/blob/master/lib/ice_cube/parsers/ical_parser.rb#L77-L78).

jkehres commented on lib/ice_cube/validations/monthly_by_set_pos.rb:
```diff
@@ -0,0 +1,75 @@
+        end_of_month = step_time.end_of_month
+
+        new_schedule = IceCube::Schedule.new(step_time - 1.month) do |s|
+          s.add_recurrence_rule(IceCube::Rule.from_hash(rule.to_hash.except(:by_set_pos, :count, :util)))
```
Did you mean `:until` instead of `:util`?

nehresma replied:
@jkehres Indeed I did!  Thanks for noticing this.

nehresma committed:
excluding until, not util

pacso commented:
I'll take a look. Am going to try to dedicate sometime this holiday season to get things tidied up and in better shape.

jkehres commented:
Trying this out for monthly frequencies with an interval > 1 seems to get stuck an infinite loop. I used the monthly test case from this PR and ran the following code, which defaults to an interval of 1, and everything works fine:

```ruby
schedule = IceCube::Schedule.from_ical "RRULE:FREQ=MONTHLY;COUNT=4;BYDAY=WE;BYSETPOS=4"
schedule.start_time = Time.new(2015, 5, 28, 12, 0, 0)
schedule.next_occurrence(Time.new(2015, 01, 01))
=> 2015-06-24 12:00:00 -0400
```

However, if I add an interval value > 1 then `next_occurrence` never returns:

```ruby
schedule = IceCube::Schedule.from_ical "RRULE:FREQ=MONTHLY;COUNT=4;BYDAY=WE;BYSETPOS=4;INTERVAL=2"
schedule.start_time = Time.new(2015, 5, 28, 12, 0, 0)
schedule.next_occurrence(Time.new(2015, 01, 01))
# method hangs
```

If I add a `puts` statement in `MonthlyBySetPos#Validation#validate` to log `step_time` and run the code above again I can see it just going infinitely into the future.

The same thing happens for yearly frequencies with an interval > 1.

jkehres commented on lib/ice_cube/time_util.rb:
```diff
@@ -1,5 +1,7 @@
+require 'date'
+require 'time'
+require 'active_support'
+require 'active_support/core_ext'
```
This introduces a new dependency for the gem on `active_support`. The `gemspec` file probably needs to be updated accordingly. A non-Rails Ruby app won't necessarily have active support available so it needs to be listed as an explicit dependency.

nehresma replied:
@jkehres Thanks, I've fixed this now.  These were a holdover from the original commit in 2016 but TimeUtil was refactored since then to no longer need ActiveSupport for time arithmetic.

jkehres commented:
I don't think this PR is sufficient to fully support BYSETPOS. It only deals with monthly and yearly frequencies. However,  playing around in the rrule.js demo [site](https://jakubroztocil.github.io/rrule/), BYSETPOS can have an effect on shorter frequencies too. For example these rule pairs each return different results with and without BYSETPOS:

```
RRULE:FREQ=WEEKLY;COUNT=15;INTERVAL=2;WKST=MO;BYDAY=MO,TU
RRULE:FREQ=WEEKLY;COUNT=15;INTERVAL=2;WKST=MO;BYDAY=MO,TU;BYSETPOS=1
```

```
RRULE:FREQ=DAILY;COUNT=15;INTERVAL=2;WKST=MO;BYHOUR=1,2
RRULE:FREQ=DAILY;COUNT=15;INTERVAL=2;WKST=MO;BYHOUR=1,2;BYSETPOS=1
```

```
RRULE:FREQ=HOURLY;COUNT=15;INTERVAL=2;WKST=MO;BYMINUTE=1,2
RRULE:FREQ=HOURLY;COUNT=15;INTERVAL=2;WKST=MO;BYMINUTE=1,2;BYSETPOS=1
```

```
RRULE:FREQ=MINUTELY;COUNT=15;INTERVAL=2;WKST=MO;BYSECOND=1,2
RRULE:FREQ=MINUTELY;COUNT=15;INTERVAL=2;WKST=MO;BYSECOND=1,2;BYSETPOS=1
```

nehresma commented:
Hey @jkehres I see what you mean regarding the infinite loop when combining interval and bysetpos.  I'm looking into it.

Also, you're right regarding the use of bysetpos and other intervals.  I'll look into adding support for those as well.

@NicolasMarlier is the original author of this work.  I've been shepherding it through bitrot issues over the years and applying feedback.

Thanks for your time evaluating this.

nehresma committed:
remove no longer needed TimeUtil active_support require

This was a holdover from the original PR back in 2016.  TimeUtil
has since been refactored to not need this, but the require was
inadvertently left.

nehresma committed:
fix interval use with bysetpos

nehresma commented:
@jkehres I've fixed the interval use for month and year frequencies with bysetpos.  I ran out of time for adding support for the other frequencies yet.  I'll add support for those after Christmas.

jkehres commented on lib/ice_cube/validations/monthly_by_set_pos.rb:
```diff
@@ -0,0 +1,76 @@
+      end
+
+      def validate(step_time, start_time)
+        start_of_month = step_time.beginning_of_month
```
The `beginning_of_month` method is not available in vanilla Ruby. It's from active support (defined [here](https://github.com/rails/rails/blob/main/activesupport/lib/active_support/core_ext/date_and_time/calculations.rb#L125)). I tried testing this PR out in a non-Rails IRB session and I get an undefined method error here because active support is not available. This PR either needs to add a dependency on the active support gem or it needs re-implement `beginning_of_month` with only vanilla Ruby. I'm not a maintainer of this gem (just an interested party that would like to help get this functionality merged) but I suspect that adding a new dependency on active support may not be something they'd be willing to accept since this gem currently has no dependencies.

jkehres replied:
Without using active support, I think this could be:

```ruby
start_of_month = IceCube::TimeUtil.build_in_zone([step_time.year, step_time.mo
nth, 1, 0, 0, 0], step_time)
```

jkehres replied:
Inspiration for this approach was taken from the existing `beginning_of_date` method (defined [here](https://github.com/ice-cube-ruby/ice_cube/blob/master/lib/ice_cube/time_util.rb#L137)).

jkehres replied:
Maybe it's even worth creating `beginning_of_xxx` and `end_of_xxx` helpers in `IceCube::TimeUtil` that are equivalent to their active support counterparts and then used them in the code here.

nehresma replied:
Nice catch.  Pushed 621bf3db42b20b38563da1cb70ca202bc6ce6f4a to resolve this.

jkehres commented on lib/ice_cube/validations/yearly_by_set_pos.rb:
```diff
@@ -0,0 +1,77 @@
+      end
+
+      def validate(step_time, start_time)
+        start_of_year = step_time.beginning_of_year
```
Ditto my comment from the previous file regarding `beginning_of_month`. The `beginning_of_year` method is only available via active support (defined [here](https://github.com/rails/rails/blob/main/activesupport/lib/active_support/core_ext/date_and_time/calculations.rb#L179)).

jkehres replied:
Without active support:

```ruby
start_of_month = IceCube::TimeUtil.build_in_zone([step_time.year, 1, 1, 0, 0, 0], step_time)
```

jkehres commented on lib/ice_cube/validations/monthly_by_set_pos.rb:
```diff
@@ -0,0 +1,76 @@
+        end_of_month = step_time.end_of_month
+
+        # Needs to start on the first day of the month
+        new_schedule = IceCube::Schedule.new(start_of_month.change(hour: step_time.hour, min: step_time.min, sec: step_time.sec)) do |s|
```
The `change` method is also from active support (defined [here](https://github.com/rails/rails/blob/main/activesupport/lib/active_support/core_ext/time/calculations.rb#L138)).

jkehres replied:
Without active support:

```ruby
IceCube::TimeUtil.build_in_zone([start_of_month.year, start_of_month.month, start_of_month.day, step_time.hour, step_time.min, step_time.sec], start_of_month)
```

jkehres commented on lib/ice_cube/validations/monthly_by_set_pos.rb:
```diff
@@ -0,0 +1,76 @@
+
+      def validate(step_time, start_time)
+        start_of_month = step_time.beginning_of_month
+        end_of_month = step_time.end_of_month
```
Without active support, this might work:

```ruby
last_day = IceCube::TimeUtil.days_in_month(step_time)
end_of_month = IceCube::TimeUtil.build_in_zone([step_time.year, step_time.mont
h, last_day, 23, 59, 59], step_time)
```

jkehres commented on lib/ice_cube/validations/yearly_by_set_pos.rb:
```diff
@@ -0,0 +1,77 @@
+
+      def validate(step_time, start_time)
+        start_of_year = step_time.beginning_of_year
+        end_of_year = step_time.end_of_year
```
Without active support:

```ruby
end_of_year = IceCube::TimeUtil.build_in_zone([step_time.year, 12, 31, 23, 59, 59], step_time)
```

nehresma committed:
remove unneeded use of activesupport for date arithmetic

nehresma committed:
support for bysetpos with freq=weekly

nehresma commented:
I've pushed bysetpos support for weekly frequencies too.  I don't think bysetpos makes sense with frequencies of daily, hourly, minutely, or secondly.  RFC 5545 makes no mention of them, and indeed all of it's examples are with monthly and weekly frequencies.  If someone can think of an example where using bysetpos with one of these other frequencies makes sense, let me know.

jkehres commented on spec/examples/to_yaml_spec.rb:
```diff
@@ -78,7 +78,7 @@ module IceCube
 
     it "should be able to make a round-trip to YAML with .day_of_year" do
-      schedule1 = Schedule.new(Time.now)
+      schedule1 = Schedule.new(Time.zone.now)
```
`Time.zone` is another active support [extension](https://github.com/rails/rails/blob/main/activesupport/lib/active_support/core_ext/time/zones.rb#L14). The stock `Time` class in Ruby has no `zone` method.

jkehres commented:
> I've pushed bysetpos support for weekly frequencies too. I don't think bysetpos makes sense with frequencies of daily, hourly, minutely, or secondly. RFC 5545 makes no mention of them, and indeed all of it's examples are with monthly and weekly frequencies. If someone can think of an example where using bysetpos with one of these other frequencies makes sense, let me know.

I believe BYSETPOS support is still needed for daily, hourly, and minutely frequencies to be compliant with the RFC, which I think should be the goal of this PR. As I mentioned in my comment [here](https://github.com/ice-cube-ruby/ice_cube/pull/449#issuecomment-1363384247), you can see BYSETPOS having an effect for these other frequencies using rrule.js - another library that implements RFC 5545 (itself a port of the python-dateutil library). If ice_cube and these other libraries all properly implement the spec, they should all give the same answer for the same RRULE. We could use that fact to aid in testing this PR to ensure we have a compliant implementation of the RFC. We could us one of these other libraries to generate a more exhaustive set of test inputs and outputs for RRULEs that use BYSETPOS and, thus, increase the number of tests for this new code along with our confidence in it.

Secondly, even though the RFC gives no examples for these other frequencies with BYSETPOS, I think you can derive that they are necessary from the details in the spec.

From the description of BYSETPOS (emphasis mine):

> The BYSETPOS rule part specifies a COMMA-separated list of values that corresponds to **the nth occurrence within the set of recurrence instances** specified by the rule.  BYSETPOS operates on **a set of recurrence instances in one interval of the recurrence rule**.

My understanding of the spec is that "interval" in this context means one unit of the frequency regardless of the INTERVAL rule part so one year for yearly, one month for monthly, etc. If there are multiple recurrences instances within one interval (e.g. year, month, etc.), then BYSETPOS has an effect and selects the nth one. Normally, there is only a single recurrence instance within an interval. The way you get more than one recurrence instance in an interval is by adding another BYxxx rule part.

Again, from the spec (emphasis mine):

> BYxxx rule parts modify the recurrence in some manner.  BYxxx rule parts for a period of time that is the same or greater than the frequency generally reduce or limit the number of occurrences of the recurrence generated.  For example, "FREQ=DAILY;BYMONTH=1" reduces the number of recurrence instances from all days (if BYMONTH rule part is not present) to all days in January. **BYxxx rule parts for a period of time less than the frequency generally increase or expand the number of occurrences of the recurrence.** For example, "FREQ=YEARLY;BYMONTH=1,2" increases the number of days within the yearly recurrence set from 1 (if BYMONTH rule part is not present) to 2.

Furthermore, this table from the spec shows how each BYxxx rule part interacts with a given frequency and whether it expands or limits the number of recurrences instances (emphasis mine):

```
   +----------+--------+--------+-------+-------+------+-------+------+
   |          |SECONDLY|MINUTELY|HOURLY |DAILY  |WEEKLY|MONTHLY|YEARLY|
   +----------+--------+--------+-------+-------+------+-------+------+
   |BYMONTH   |Limit   |Limit   |Limit  |Limit  |Limit |Limit  |Expand|
   +----------+--------+--------+-------+-------+------+-------+------+
   |BYWEEKNO  |N/A     |N/A     |N/A    |N/A    |N/A   |N/A    |Expand|
   +----------+--------+--------+-------+-------+------+-------+------+
   |BYYEARDAY |Limit   |Limit   |Limit  |N/A    |N/A   |N/A    |Expand|
   +----------+--------+--------+-------+-------+------+-------+------+
   |BYMONTHDAY|Limit   |Limit   |Limit  |Limit  |N/A   |Expand |Expand|
   +----------+--------+--------+-------+-------+------+-------+------+
   |BYDAY     |Limit   |Limit   |Limit  |Limit  |Expand|Note 1 |Note 2|
   +----------+--------+--------+-------+-------+------+-------+------+
   |BYHOUR    |Limit   |Limit   |Limit  |Expand |Expand|Expand |Expand|
   +----------+--------+--------+-------+-------+------+-------+------+
   |BYMINUTE  |Limit   |Limit   |Expand |Expand |Expand|Expand |Expand|
   +----------+--------+--------+-------+-------+------+-------+------+
   |BYSECOND  |Limit   |Expand  |Expand |Expand |Expand|Expand |Expand|
   +----------+--------+--------+-------+-------+------+-------+------+
   |BYSETPOS  |Limit   |Limit   |Limit  |Limit  |Limit |Limit  |Limit |
   +----------+--------+--------+-------+-------+------+-------+------+
```

For secondly, you can see that no BYxxx rule part will expand the set of recurrence instances so BYSETPOS won't have an effect. However, for minutely, hourly, and daily frequencies you can see that some BYxxx rule parts will expand the set of recurrences, which would allow BYSETPOS to select the nth one in an interval and thus effect the final set of recurrence instances.

jkehres commented on spec/examples/by_set_pos_spec.rb:
```diff
@@ -0,0 +1,103 @@
+  describe WeeklyRule, 'BYSETPOS' do
+    it 'should behave correctly' do
+      # Weekly on Monday, Wednesday, and Friday with the week starting on Wednesday, the last day of the set
+      schedule = IceCube::Schedule.from_ical("RRULE:FREQ=WEEKLY;COUNT=4;WKST=WE;BYDAY=MO,WE,FR;BYSETPOS=-1")
```
For each of the frequencies, it would be nice to have the following test cases:

- With BYxxx that expands the set of recurrence instances...
  - BYSETPOS with a positive value
  - BYSETPOS with a negative value
  - BYSETPOS with multiple positive values
  - BYSETPOS with multiple negative values
  - BYSETPOS with multiple values both positive and negative
  - BYSETPOS with multiple repeated values - spec does not define behaviour; rrule.js does some weird things and outputs the same occurrence multiple times (bug or maybe just undefined behaviour?)
  - BYSETPOS with a value that matches in some intervals but not others (e.g. selecting the 5th Monday of every month since some months will have only 4 Mondays)
  - BYSETPOS with a positive value beyond than the number of recurrence instances in any interval (e.g. the 10th Monday of every month); rrule.js returns an empty set of occurrences
  - BYSETPOS with a negative value bigger than the number of recurrence instances in any set (e.g. the -10th Monday of every month); rrule.js returns a non-empty set of occurrences, which is weird (bug or maybe just undefined behaviour?)
- BYSETPOS with BYxxx that limits the set of recurrence instances instead of expanding it
- BYSETPOS with no BYxxx - spec says BYSETPOS "MUST only be used in conjunction with another BYxxx rule part" so not sure if this should be an exception or an empty set of occurrences returned; rrule.js does the latter
- BYSETPOS with multiple BYxxx rules parts that expand the recurrence instances
- BYSETPOS with multiple BYxxx rules parts that both expand and limit the recurrence instances
- BYSETPOS with a value outside the allowed range (covered in the `from_ical` tests but can you also construct a schedule using BYSETPOS without `from_ical`? if not, should that be a feature of this PR?)

jkehres commented on lib/ice_cube/parsers/ical_parser.rb:
```diff
@@ -75,7 +75,7 @@ def self.rule_from_ical(ical)
           validations[:day_of_year] = value.split(",").map(&:to_i)
         when "BYSETPOS"
-          # noop
+          params[:validations][:by_set_pos] = value.split(',').collect(&:to_i)
```
Also, existing code uses double quotes instead of single quotes for string in `split` and `collect` instead of `map`. Picky but just for consistency.

jkehres commented on lib/ice_cube/time_util.rb:
```diff
@@ -1,5 +1,5 @@
-require "date"
-require "time"
+require 'date'
+require 'time'
```
Changing the type of quotes seems unnecessary. Double quotes are consistent with the existing style.

jkehres commented on lib/ice_cube/validations/weekly_by_set_pos.rb:
```diff
@@ -0,0 +1,91 @@
+          ], step_time
+        )
+        new_schedule = IceCube::Schedule.new(start_of_week_adjusted) do |s|
+          s.add_recurrence_rule(IceCube::Rule.from_hash(rule.to_hash.except(:by_set_pos, :count, :until)))
```
`except` is another active support method:

https://github.com/rails/rails/blob/v7.0.4/activesupport/lib/active_support/core_ext/hash/except.rb#L12

jkehres commented on lib/ice_cube/validations/monthly_by_set_pos.rb:
```diff
@@ -0,0 +1,77 @@
+
+        # Needs to start on the first day of the month
+        new_schedule = IceCube::Schedule.new(IceCube::TimeUtil.build_in_zone([start_of_month.year, start_of_month.month, start_of_month.day, step_time.hour, step_time.min, step_time.sec], start_of_month)) do |s|
+          s.add_recurrence_rule(IceCube::Rule.from_hash(rule.to_hash.except(:by_set_pos, :count, :until)))
```
Ditto my [comment](https://github.com/ice-cube-ruby/ice_cube/pull/449/files#r1067495750) about `except`.

jkehres commented on lib/ice_cube/validations/yearly_by_set_pos.rb:
```diff
@@ -0,0 +1,77 @@
+
+        # Needs to start on the first day of the year
+        new_schedule = IceCube::Schedule.new(IceCube::TimeUtil.build_in_zone([start_of_year.year, start_of_year.month, start_of_year.day, step_time.hour, step_time.min, step_time.sec], start_of_year)) do |s|
+          s.add_recurrence_rule(IceCube::Rule.from_hash(rule.to_hash.except(:by_set_pos, :count, :until)))
```
Ditto my [comment](https://github.com/ice-cube-ruby/ice_cube/pull/449/files#r1067495750) about `except`.

jkehres commented:
Using BYSECOND with MONTHLY does not work. I tried this:

```ruby
ical = "RRULE:FREQ=MONTHLY;COUNT=4;BYSECOND=1,2,3,4;BYSETPOS=2"
schedule = IceCube::Schedule.from_ical(ical)
schedule.start_time = Time.new(2015, 5, 28, 12, 0, 0)
schedule.occurrences_between(Time.new(2015, 01, 01), Time.new(2017, 01, 01))
```

and got an empty array. I expected to get:

```
2015-05-28 12:00:02
2015-06-28 12:00:02
2015-07-28 12:00:02
2015-08-28 12:00:02
```

I suspect there is an assumption in the code that you will only use BYDAY with MONTHLY and not other BYxxx rule parts.

nehresma committed:
support for parsing rrules from ical that are very long and wrap

nehresma committed:
dont require the wrapped line to be the last the ical string

rgalanakis commented:
First I want to thank @nehresma @jkehres and others for all the work on this.

> I believe BYSETPOS support is still needed for daily, hourly, and minutely frequencies to be compliant with the RFC, which I think should be the goal of this PR.

The current code already doesn't implement the RFC, and it fails to do so silently (just spent over an hour finding the this as the source of a bug). I think partial RFC compliance is better than none. And the users who use daily/hourly/minutely frequencies would continue to see the current unimplemented behavior. I know I wouldn't be alone in appreciating getting this work merged.

pacso commented:
Thanks for the review @jkehres. Can we get the tests running/passing?

jorgemanrubia commented:
Thank you so much for your work here. I'd love if we can get this merged. Apple Calendar feeds use `BYSETPOS` when selecting a day of the month. And there is no direct equivalent for the negative values (starting from the end). So this would be super handy.

pacso commented:
@jorgemanrubia - Since nothing has moved on this PR for a long time ... it might be easiest if you can fork it, get the tests passing and then we can move forwards? I have struggled for time recently otherwise would have done this myself.

Skulli commented:
Monthly/Yearly would be more than enough for me.

nehresma commented on lib/ice_cube/validations/monthly_by_set_pos.rb:
```diff
@@ -0,0 +1,85 @@
+
+
+        new_schedule = IceCube::Schedule.new(TimeUtil.previous_month(step_time)) do |s|
+          s.add_recurrence_rule IceCube::Rule.from_hash(rule.to_hash.reject{|k, v| [:by_set_pos, :count, :until].include? k})
```
Yes but ActiveSupport isn't a runtime requirement of this gem.

nehresma committed:
nitpick fixes - use map and double quotes for consistency

nehresma committed:
do not rely on ActiveSupport-only helper methods

nehresma committed:
fix BYSETPOS serialization

nehresma committed:
fix SETBYPOS with non BYDAY expansions

nehresma committed:
expand BYSETPOS spec coverage

- weekly: positive/multi/mixed positions, BYHOUR expansion
- monthly: negative positions, BYMONTHDAY, BYMINUTE
- yearly: multiple positive/negative positions

nehresma committed:
adding BYSETPOS validations for daily/hourly/minutely

nehresma committed:
add verification that bysetpos runs after other byXXX filters

nehresma committed:
BYSETPOS interval boundaries specs

nehresma committed:
dd BYSETPOS ordering specs

nehresma committed:
add BYSETPOS anchor and interval specs

nehresma committed:
refactor and create BYSETPOS helper for interval bounds

nehresma committed:
adding more comprehensive BYSETPOS specs

This covers repeated values, out-of-range, and UNTIL cases

nehresma committed:
updating readme with bysetpos support

nehresma committed:
adding BYSETPOS to_ical spec coverage

nehresma committed:
adding SECONDLY BYSETPOS support and specs

nehresma committed:
adding BYYEARDAY BYSETPOS specs

nehresma committed:
Gemfile adjustments for Ruby stdlib changes

nehresma committed:
adding support for more versions of ActiveSupport

nehresma commented:
Hmm, yes you are right -- I would expect the 11th to be the first event.  I haven't tested this particular event with this branch of ice_cube yet.  I'll take a closer look at this, fix it, and add test coverage for this to verify it.  Thanks!!!

nehresma committed:
Fix BYSETPOS to count positions from interval start per RFC 5545

BYSETPOS was anchoring candidate enumeration to DTSTART instead of the
interval boundary, causing positions to be miscounted. This skipped the
first occurrence when DTSTART itself matched a BYSETPOS position (e.g.,
starting on the 2nd Tuesday when selecting BYSETPOS=2).

The fix anchors the temporary schedule to the interval start based on
which BYxxx components expand the candidate set (day, month, hour, etc.).

nehresma committed:
Fix flaky YAML round-trip test that failed across DST boundaries

Use UTC instead of Time.now to avoid DST offset mismatches when YAML
serialization loses timezone info and preserves only the numeric offset.

rgalanakis commented:
Thanks, confirmed working on my end. Appreciate the quick response as it saved some git headaches (I stupidly took a dependency directly on your fork)

nehresma commented:
@pacso I can definitely address the linting issues in the code that was changed in this pull request, but I'll leave the code that wasn't touched here alone.  It'll make merging easier.

nehresma committed:
linting fixes

nehresma committed:
CHANGELOG entry

pacso committed:
Merge branch 'master' into bysetpos

pacso committed:
Fix lint errors

pacso committed:
Add missing round-trip YAML tests for BYSETPOS

pacso committed:
Create time objects correctly

pacso committed:
Add edge cases for BYSETPOS

pacso commented:
Thanks everybody for all of your hard work! I'll see what else is ready to merge this week and then get a new release out.

AntoineGirard commented:
Thanks all for your work on this feature 🙇🏻

It’s a pleasure to discover that we need to support BYSETOPS this week, and then you merge this long work last week.

Long living open source!

nehresma commented:
@pacso any chance you could cut a release to rubygems.org?

pacso commented:
Yes, I will do. I just wanted to see whether there were any other PRs that ought to also be included before I did it.
